[MCCAS] Split DWARF Abbrev section

felipepiovezan opened the pull request:
With this commit, the MCCAS backend now splits the __debug_abbrev
section into multiple DebugAbbrevRef CAS objects.

In order to split the abbreviation section, we require information from
the __debug_info compilation units; in particular, a list of
abbreviation offsets is used to derive information about _where_ the
split points are. For example, if the compilation units have
abbreviation offsets {10, 20, 0}, we deduce that the abbreviation
section can be split according to the intervals [0, 10), [10, 20) and
[20, size(__debug_abbrev)).

This commit does not link the CURef to its corresponding AbbrevRef with
a CAS edge. This is ok for now: because we don't zero the Abbrev offset
in CURef, the relationship between the two nodes is preserved
indirectly.

felipepiovezan committed:
[MCCAS] Split DWARF Abbrev section

With this commit, the MCCAS backend now splits the __debug_abbrev
section into multiple DebugAbbrevRef CAS objects.

In order to split the abbreviation section, we require information from
the __debug_info compilation units; in particular, a list of
abbreviation offsets is used to derive information about _where_ the
split points are. For example, if the compilation units have
abbreviation offsets {10, 20, 0}, we deduce that the abbreviation
section can be split according to the intervals [0, 10), [10, 20) and
[20, size(__debug_abbrev)).

This commit does not link the CURef to its corresponding AbbrevRef with
a CAS edge. This is ok for now: because we don't zero the Abbrev offset
in CURef, the relationship between the two nodes is preserved
indirectly.

felipepiovezan commented on llvm/include/llvm/MC/CAS/MCCASObjectV1.h:
```diff
@@ -302,6 +302,12 @@ struct DwarfSectionsCache {
+  MCSection *Abbrev;
+};
+
+struct AbbrevAndDebugSplit {
```
This may seem overkill now, but later we will also have a third member: the CAS object representing the abbrev offsets node

felipepiovezan committed:
[MCCAS] Introduce DebugAbbrevOffsetsRef CAS object

When splitting the DebugInfo section, a new CAS object is nested under
the section: the DebugAbbrevOffsetsRef object. It contains an encoded
list of abbreviation offsets in the section.

This commit introduces an adaptor class for this CAS object, which
consolidates the entire logic of encoding/decoding the offsets so that
this logic can be reused with other tools, e.g. the cas-dump tool.

felipepiovezan commented:
Since this hasn't been reviewed yet, I've added a second commit to this PR, introducing the new abbreviation offset node

adrian-prantl commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -985,6 +987,57 @@ getSizeFromDwarfHeaderAndSkip(BinaryStreamReader &Reader) {
   return Size;
 }
 
+/// Reads and returns the abbreviation offset field of a dwarf Compilation Unit
```
Nit: we should pick one spelling for Dwarf and stick to it ;-)

felipepiovezan replied:
Sigh, yeah :(
For some reason spelling Dwarf is very mood dependent...

adrian-prantl commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1033,21 +1086,92 @@ splitDebugInfoSectionData(MutableArrayRef<char> SectionData,
+
+  // Sorted container of offsets such that, for any iterator `it`, the interval
+  // [*it, *(it+1)] describes where to find one Abbreviation contribution.
+  SmallSet<size_t, 8> AbbrevOffsets;
```
Just curious: where does that 8 come from and why not a DenseSet?

felipepiovezan replied:
The 8 is something that should have been cleaned up. More importantly though, both SmallSet and DenseSet are wrong here: because we _require_ sorted iteration order, the correct choice is either a sorted + uniqued vector or a std::set. I'll update this

adrian-prantl commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1033,21 +1086,92 @@ splitDebugInfoSectionData(MutableArrayRef<char> SectionData,
+
+  size_t StartOffset = *AbbrevOffsets.begin();
+  if (StartOffset != 0)
+    return createStringError(inconvertibleErrorCode(), "Invalid Dwarf");
```
is there a more targeted error message we could use here? "Expected nonzero start offset"?

adrian-prantl commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1033,21 +1086,92 @@ splitDebugInfoSectionData(MutableArrayRef<char> SectionData,
+  SmallVector<DebugAbbrevRef> AbbrevRefs;
+  for (size_t EndOffset : llvm::drop_begin(AbbrevOffsets)) {
+    StringRef AbbrevData =
+        toStringRef(*FullAbbrevData).slice(StartOffset, EndOffset);
```
Nice.

adrian-prantl commented on llvm/test/DebugInfo/CAS/AArch64/debug_abbrev_split.ll:
```diff
@@ -0,0 +1,51 @@
+; RUN: llc -O0 --filetype=obj --cas-backend --cas=%t.casdb --mccas-verify -o /dev/null %s
```
os `/dev/null` something lit specifically knows about? I.e., would this test work on Windows?

adrian-prantl replied:
otherwise `%t.bin`  be a compatible alternative.

felipepiovezan replied:
Good catch!

adrian-prantl commented on llvm/tools/llvm-cas-dump/llvm-cas-dump.cpp:
```diff
@@ -26,6 +26,8 @@ cl::list<std::string> InputStrings(cl::Positional,
 cl::opt<bool> DwarfSectionsOnly("dwarf-sections-only",
                                 cl::desc("Only print dwarf related sections"));
+cl::opt<bool> DebugAbbrevOffsets("debug-abbrev-offsets",
+                                 cl::desc("Only print dwarf related sections"));
```
copy&paste error.

adrian-prantl reviewed (commented):
Nice!

felipepiovezan reviewed (commented):
Addressed review comments, rebased

felipepiovezan commented:
Rebased and fixed conflicts with the new capabilities introduced in the cas-dump tool.
@adrian-prantl please double check my changes to the dwarf dump code (second commit) :)

rastogishubham commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1062,6 +1166,16 @@ Expected<SmallVector<DebugInfoCURef>> MCCASBuilder::splitDebugInfoSection() {
     return SplitCUData.takeError();
 
+  Expected<SmallVector<size_t>> AbbrevOffsets =
+      getAbbrevSectionOffsets(*SplitCUData, Asm.getBackend().Endian);
```
I think it would be more efficient if you could just return the abbrev offsets in `splitDebugInfoSectionData` rather than split them and then once again create another `BinaryStreamReader` to read the data and get the abbrev offsets

adrian-prantl commented on llvm/include/llvm/MC/CAS/MCCASObjectV1.h:
```diff
@@ -302,6 +302,29 @@ struct DwarfSectionsCache {
+};
+
+// Helper class to allow reusing the logic of encoding/decoding Abbreviation
+// Offsets.
```
///

adrian-prantl commented on llvm/tools/llvm-cas-dump/CASDWARFObject.cpp:
```diff
@@ -109,20 +79,9 @@ Error CASDWARFObject::discoverDwarfSections(MCObjectProxy MCObj) {
-  }
-
+  if (DebugAbbrevRef::Cast(MCObj))
+    append_range(DebugAbbrevSection, MCObj.getData());
```
Much better!

adrian-prantl commented on llvm/tools/llvm-cas-dump/CASDWARFObject.cpp:
```diff
@@ -109,20 +79,9 @@ Error CASDWARFObject::discoverDwarfSections(MCObjectProxy MCObj) {
-    }
-  }
-
+  if (DebugAbbrevRef::Cast(MCObj))
```
else if

felipepiovezan replied:
Ahh I want a `TypeSwitch` so bad 😭

adrian-prantl commented on llvm/tools/llvm-cas-dump/CASDWARFObject.cpp:
```diff
@@ -109,20 +79,9 @@ Error CASDWARFObject::discoverDwarfSections(MCObjectProxy MCObj) {
-
+  if (DebugAbbrevRef::Cast(MCObj))
+    append_range(DebugAbbrevSection, MCObj.getData());
   if (DebugStrRef::Cast(MCObj)) {
```
else if

adrian-prantl reviewed (approved):
Minor nitpicks inside, otherwise looking good!

rastogishubham commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1062,6 +1166,16 @@ Expected<SmallVector<DebugInfoCURef>> MCCASBuilder::splitDebugInfoSection() {
+    return AbbrevOffsets.takeError();
+
+  Expected<SmallVector<DebugAbbrevRef>> AbbrevRefs =
+      splitAbbrevSection(*AbbrevOffsets);
```
There is a small inconsistency here, and it is a stylistic thing, but in the function, `splitDebugInfoAndAbbrevSections` you pass the `DebugInfoData` to `splitDebugInfoSectionData` where the `DebugInfoData` is the data in the debug info section after merging the fragments, but `splitAbbrevSection` takes just the `AbbrevOffsets` and does not take the merged abbrev data as an argument, can we also move the call to `mergeMCFragmentContents` in `splitDebugInfoSectionData` and rename it to `splitDebugInfoSection`, this way, `splitDebugInfoSection` could also return a vector of `AbbrevOffsets`. Maybe we can use `splitDebugInfoSectionAndGetAbbrevOffsets` as the name?

felipepiovezan replied:
>, can we also move the call to mergeMCFragmentContents in splitDebugInfoSectionData

We can and I will do so, but we are playing with fire here: this function will return a sequence of MutableArrayRefs into the merged data vector. If said vector is created _inside_ the function, how do we guarantee the MutableArrayRefs are not invalidated?
The function needs to _move_ the merged data vector out of it as well. But this is not enough: iterators can be invalidated across moves for containers using the "Small Size optimization". So we also rely on the small size being 0.

Anyhow, I would normally have pushed back against this suggestion, but to get the review moving faster I'll go ahead and implement this change

felipepiovezan replied:
> So we also rely on the small size being 0.

Actually, I think even this is not guaranteed to be enough...

felipepiovezan replied:
I've made the changes to make the abbreviation offset extraction single pass, but for now I'd advise against moving the merged vector around, it's just too fragile to do what you suggested because of memory management

rastogishubham replied:
I think that is fair, I just had a suggestion, and your reasoning is valid!

rastogishubham commented on llvm/lib/MC/MCCASObjectV1.cpp:
```diff
@@ -1045,9 +1098,60 @@ Error MCCASBuilder::createDebugInfoSection(ArrayRef<DebugInfoCURef> CURefs) {
+MCCASBuilder::splitAbbrevSection(ArrayRef<size_t> AbbrevOffsetVec) {
+  // If we are here, the Abbrev section _must_ exist.
+  if (!DwarfSections.Abbrev)
+    return createStringError(inconvertibleErrorCode(), "Invalid Dwarf");
```
Maybe we want to be a bit more verbose and say "Invalid Dwarf, no Debug Abbrev Section present"?

rastogishubham reviewed (commented):
I had a few thoughts, let me know what you think!

felipepiovezan commented:
@swift-ci please test

felipepiovezan commented:
Rebased

felipepiovezan commented:
@swift-ci please test

rastogishubham reviewed (approved):
LGTM!

felipepiovezan commented:
Rebased

@swift-ci please test
